Skip nvm aliases in Travis CI rule

sonicdoe opened the pull request:
When running clinton against a package with an `engines` field and `node` in `.travis.yml` it crashes with `TypeError: Invalid Version: node.0.0`. A similar error was already reported in #46.

With this change it now skips all nvm aliases in `isSupported()`. It still crashes when using the `unstable` alias but I suggest resolving this by adding `unstable` to the list of deprecated versions.

SamVerschueren commented:
And to which version is `node` mapped? Latest? As you can see in [this line](https://github.com/SamVerschueren/clinton/blob/master/rules/travis.js#L5), `stable` for instance is marked as deprecated. So actually I'm more in favour of marking these versions as not supported because it's not explicit about which versions it tests.

// @sindresorhus

sonicdoe commented:
`stable` was deprecated because it “only truly applies to node v0.12 and earlier”. `node` always maps to the latest version of Node.js.

I am a bit more in favor of ignoring the `node` alias as opposed to marking it as an error because it is useful to always test against the latest Node.js version. However, I can understand if users should be urged to keep their `.travis.yml` up to date, instead.

sindresorhus commented:
I agree `node` should just be ignored. It can be useful sometimes when you just want the latest Node.js version. [Example.](https://github.com/sindresorhus/refined-github/blob/1711eaac42d84f1ab304b73e6a606dde1df82175/.travis.yml#L3)

SamVerschueren commented:
Good point. @sonicdoe do you happen to have a reference of `unstable` being part of Travis? Can't seem to find it.

Let's ignore `node`. which seems it was [already ignored](https://github.com/SamVerschueren/clinton/blob/master/rules/travis.js#L85) but somehow fails on some other scenarios which aren't tested. All the other should be moved to the deprecated versions array, which only is the `unstable` version. But I want proof first of that alias to exist.

In order to ignore `node` decently, I'd suggest just using `filter` on the [versions array](https://github.com/SamVerschueren/clinton/blob/master/rules/travis.js#L77) so we don't have to take that one into account further down in the code.

sonicdoe commented:
[Travis CI uses nvm](https://docs.travis-ci.com/user/languages/javascript-with-nodejs#Choosing-Node-versions-to-test-against):

> These values are passed on to `nvm`; newer releases not shown above may be used if `nvm` recognizes them.

nvm itself lists `node`, `iojs`, `stable`, and `unstable` as [special default aliases](https://github.com/creationix/nvm#usage). I have also just realized that there are [LTS aliases](https://github.com/creationix/nvm#long-term-support). Given that these are similar to `node`, I have added them to the ignore array, too.

I have updated this pull request and opened another one [deprecating `unstable`](https://github.com/SamVerschueren/clinton/pull/65).

sonicdoe commented on rules/travis.js:
```diff
@@ -74,15 +75,23 @@ module.exports = ctx => {
+				}
+			}
+
+			versions = versions.filter(v => DEPRECATED_VERSIONS.indexOf(v) === -1);
```
I could have merged this filter with the above for…of loop, however, I think that the code is more readable this way and the performance impact is negligible.

SamVerschueren replied:
This is good!

SamVerschueren replied:
I'm looking back at the code and wondering why we can't write the above for loop like what we had before.

```js
for (const version of versions) {
	if (DEPRECATED_VERSIONS.indexOf(version) !== -1) {
		ctx.report({
			message: `Version \`${version}\` is deprecated.`,
			file
		});
	} else if (engine && !semver.satisfies(normalize(version), engine)) {
		ctx.report({
			message: `Unsupported version \`${version}\` is being tested.`,
			fix: fixers.unsupported(version),
			file
		});
	}
}
```

I believe this does the same thing as your code with 2 for-loops without having to filter out the deprecated versions.

sonicdoe replied:
We would then have to filter out the deprecated versions again in the [`if (engine)` block](https://github.com/SamVerschueren/clinton/blob/v0.11.0/rules/travis.js#L94-L104) further down. [Like you said](https://github.com/SamVerschueren/clinton/pull/64#issuecomment-276000160):

> In order to ignore node decently, I'd suggest just using filter on the versions array so we don't have to take that one into account further down in the code.

SamVerschueren replied:
Oh yes, you're right. Thanks for clarifying.

Although, maybe it might be a better idea to merge the two for..of loops and filter the `versions` array again after that for..of loop right before we enter the `if (engine)` block. I think it's just weird that we would do it in two loops while we can easily do it in one loop.

sonicdoe replied:
What do you think about merging them all into one filter?

```js
let versions = Array.isArray(travis.node_js) ? travis.node_js : [travis.node_js];

versions = versions.filter(version => {
  if (IGNORED_VERSIONS.indexOf(version) !== -1) {
    return false;
  } else if (DEPRECATED_VERSIONS.indexOf(version) !== -1) {
    ctx.report({
      message: `Version \`${version}\` is deprecated.`,
      file
    });

    return false;
  } else if (engine && !semver.satisfies(normalize(version), engine)) {
    ctx.report({
      message: `Unsupported version \`${version}\` is being tested.`,
      fix: fixers.unsupported(version),
      file
    });
  }

  return true;
});
```

This is what I initially tried to avoid but, on second thought, it might not be so bad.

SamVerschueren replied:
Looks good to me!

SamVerschueren commented:
> I have updated this pull request and opened another one deprecating unstable.

Just deprecate `unstable` in this PR instead of a new one.

SamVerschueren commented on rules/travis.js:
```diff
@@ -74,15 +75,23 @@ module.exports = ctx => {
 
-			const versions = Array.isArray(travis.node_js) ? travis.node_js : [travis.node_js];
+			let versions = Array.isArray(travis.node_js) ? travis.node_js : [travis.node_js];
+
```
Remove the empty line

sonicdoe commented:
Done.

SamVerschueren commented:
Awesome, great work @sonicdoe! I created a follow-up issue #66 for an extra improvement on this rule.
